File: runtime/typescript/packages/anthropic/src/processor.ts
/**
 * Anthropic processor — extracts clean results from Anthropic Messages API responses.
 *
 * Handles:
 * - Text content from `content[]` blocks
 * - Tool use blocks → ToolCall objects
 * - Streaming responses (content_block_delta events)
 * - Structured output (JSON parse when outputs are present)
 *
 * @module
 */

import type { Prompty } from "@prompty/core";
import type { Processor } from "@prompty/core";
import type { ToolCall } from "@prompty/core";
import {
  ErrorChunk,
  InvocationUsage,
  StreamChunk,
  TextChunk,
  ThinkingChunk,
  ToolChunk,
  UsageChunk,
  traceSpan,
} from "@prompty/core";
import { createStructuredResult } from "@prompty/core";

export class AnthropicProcessor implements Processor {
  async process(agent: Prompty, response: unknown): Promise<unknown> {
    return traceSpan("AnthropicProcessor", async (emit) => {
      emit("signature", "prompty.anthropic.processor.AnthropicProcessor.invoke");
      emit("inputs", { data: response });
      const result = processResponse(agent, response);
      // Don't emit result for streaming — it's a generator, not a value
      if (!isAsyncIterable(response)) {
        emit("result", result);
      }
      return result;
    });
  }

  processStream(response: AsyncIterable<unknown>): AsyncIterable<StreamChunk> {
    return processStream(response);
  }
}

/**
 * Extract clean content from an Anthropic Messages API response.
 */
export function processResponse(agent: Prompty, response: unknown): unknown {
  if (typeof response !== "object" || response === null) return response;

  // Streaming response — return content-extracting async generator
  if (isAsyncIterable(response)) {
    return legacyStreamGenerator(processStream(response));
  }

  const r = response as Record<string, unknown>;

  // Anthropic Messages response — has `content` array and `role`
  if (Array.isArray(r.content) && r.role === "assistant") {
    return processMessages(agent, r);
  }

  return response;
}

// ---------------------------------------------------------------------------
// Streaming
// ---------------------------------------------------------------------------

/** Type guard for async iterables (PromptyStream or raw SDK stream). */
function isAsyncIterable(value: unknown): value is AsyncIterable<unknown> {
  return (
    typeof value === "object" &&
    value !== null &&
    Symbol.asyncIterator in value
  );
}

/**
 * Yield content chunks from an Anthropic streaming response.
 *
 * Anthropic streaming events include:
 * - `content_block_delta` with `delta.type === "text_delta"` → yield text
 * - `content_block_start` with `content_block.type === "tool_use"` → accumulate tool call
 * - `input_json` deltas for tool arguments
 * - `message_stop` → end of stream
 *
 * Tool calls are accumulated and yielded at the end of the stream.
 */
export async function* processStream(
  response: AsyncIterable<unknown>,
): AsyncGenerator<StreamChunk> {
  const toolCallAcc: Map<
    number,
    { id: string; name: string; arguments: string }
  > = new Map();
  let inputTokens: number | undefined;
  let outputTokens: number | undefined;

  try {
    for await (const event of response) {
      const e = event as Record<string, unknown>;
      const eventType = e.type as string | undefined;

      if (eventType === "message_start") {
        const message = e.message as Record<string, unknown> | undefined;
        const usage = message?.usage as Record<string, unknown> | undefined;
        inputTokens = numberValue(usage?.input_tokens) ?? inputTokens;
      } else if (eventType === "message_delta") {
        const usage = e.usage as Record<string, unknown> | undefined;
        outputTokens = numberValue(usage?.output_tokens) ?? outputTokens;
      } else if (eventType === "content_block_delta") {
        const delta = e.delta as Record<string, unknown> | undefined;
        if (!delta) continue;

        if (delta.type === "text_delta" && typeof delta.text === "string" && delta.text) {
          yield new TextChunk({ value: delta.text });
        } else if (delta.type === "thinking_delta" && typeof delta.thinking === "string" && delta.thinking) {
          yield new ThinkingChunk({ value: delta.thinking });
        } else if (delta.type === "input_json_delta") {
          const idx = typeof e.index === "number" ? e.index : 0;
          const acc = toolCallAcc.get(idx);
          if (acc && typeof delta.partial_json === "string") {
            acc.arguments += delta.partial_json;
          }
        }
      } else if (eventType === "content_block_start") {
        const block = e.content_block as Record<string, unknown> | undefined;
        if (block?.type === "tool_use") {
          const idx = typeof e.index === "number" ? e.index : 0;
          toolCallAcc.set(idx, {
            id: stringValue(block.id),
            name: stringValue(block.name),
            arguments: "",
          });
        }
      } else if (eventType === "error") {
        const error = e.error as Record<string, unknown> | undefined;
        yield new ErrorChunk({
          message: stringValue(error?.message) || "Anthropic stream failed",
        });
        return;
      }
    }
  } catch (error) {
    yield new ErrorChunk({
      message: error instanceof Error ? error.message : String(error),
    });
    return;
  }

  const sortedIndices = [...toolCallAcc.keys()].sort((a, b) => a - b);
  for (const idx of sortedIndices) {
    const tc = toolCallAcc.get(idx)!;
    yield ToolChunk.load({ kind: "tool", toolCall: tc });
  }
  if (inputTokens !== undefined || outputTokens !== undefined) {
    const input = inputTokens ?? 0;
    const output = outputTokens ?? 0;
    yield new UsageChunk({
      usage: new InvocationUsage({
        inputTokens: input,
        outputTokens: output,
        totalTokens: input + output,
      }),
    });
  }
}

async function* legacyStreamGenerator(
  chunks: AsyncIterable<StreamChunk>,
): AsyncGenerator<string | ToolCall> {
  for await (const chunk of chunks) {
    if (chunk instanceof TextChunk) {
      yield chunk.value;
    } else if (chunk instanceof ToolChunk) {
      yield chunk.toolCall;
    } else if (chunk instanceof ErrorChunk) {
      throw new Error(chunk.message);
    }
  }
}

function stringValue(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function numberValue(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

// ---------------------------------------------------------------------------
// Non-streaming response processing
// ---------------------------------------------------------------------------

/**
 * Process an Anthropic Messages API response.
 *
 * Response shape:
 * ```json
 * {
 *   "role": "assistant",
 *   "content": [
 *     { "type": "text", "text": "..." },
 *     { "type": "tool_use", "id": "...", "name": "...", "input": {...} }
 *   ],
 *   "stop_reason": "end_turn" | "tool_use" | "max_tokens" | "stop_sequence"
 * }
 * ```
 */
function processMessages(
  agent: Prompty,
  response: Record<string, unknown>,
): unknown {
  const content = response.content as Record<string, unknown>[];
  if (!content || content.length === 0) return null;

  // Check for tool_use blocks
  const toolCalls: ToolCall[] = [];
  const textParts: string[] = [];

  for (const block of content) {
    if (block.type === "tool_use") {
      toolCalls.push({
        id: block.id as string,
        name: block.name as string,
        arguments:
          typeof block.input === "string"
            ? (block.input as string)
            : JSON.stringify(block.input),
      });
    } else if (block.type === "text") {
      textParts.push(block.text as string);
    }
  }

  // If tool calls present, return them (pipeline handles the loop)
  if (toolCalls.length > 0) {
    return toolCalls;
  }

  // Text content
  const text = textParts.join("");
  if (!text) return null;

  // Structured output — JSON parse when outputs schema exists
  if (agent.outputs && agent.outputs.length > 0) {
    try {
      return createStructuredResult(JSON.parse(text) as Record<string, unknown>, text);
    } catch {
      return text;
    }
  }

  return text;
}

File: runtime/python/prompty/prompty/providers/anthropic/processor.py
"""Anthropic processor — extracts clean results from Anthropic Messages API responses.

Handles:
- Text content from ``content[]`` blocks
- Tool use blocks → ToolCall objects
- Streaming responses (content_block_delta events)
- Structured output (JSON parse when outputSchema present)

Registered as ``anthropic`` in ``prompty.processors``.
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator, Iterator
from typing import Any

from ...tracing.tracer import trace

__all__ = ["AnthropicProcessor"]


class AnthropicProcessor:
    """Processor for Anthropic Messages API responses."""

    @trace
    def process(self, agent: Any, data: Any) -> Any:
        return _process_response(agent, data)

    @trace
    async def process_async(self, agent: Any, data: Any) -> Any:
        return _process_response(agent, data)


def _process_response(agent: Any, response: Any) -> Any:
    """Extract clean content from an Anthropic Messages API response."""
    # Check for Anthropic Messages response shape first (has content + role).
    # This must come before the Iterator check because mock objects
    # and some SDK types may satisfy Iterator checks.
    content = None
    role = None
    stop_reason = None

    if isinstance(response, dict):
        content = response.get("content")
        role = response.get("role")
        stop_reason = response.get("stop_reason")
    elif hasattr(response, "content") and hasattr(response, "role"):
        content = getattr(response, "content", None)
        role = getattr(response, "role", None)
        stop_reason = getattr(response, "stop_reason", None)

    if content is not None and role == "assistant":
        return _process_messages(agent, content, stop_reason)

    # Streaming — sync or async iterator (raw SDK stream, not a response object)
    if isinstance(response, Iterator):
        return _stream_generator(response)
    if isinstance(response, AsyncIterator):
        return _async_stream_generator(response)

    return response


# ---------------------------------------------------------------------------
# Non-streaming response processing
# ---------------------------------------------------------------------------


def _process_messages(agent: Any, content: list[Any], stop_reason: str | None) -> Any:
    """Process content blocks from an Anthropic response."""
    from ..openai.processor import ToolCall

    tool_calls: list[ToolCall] = []
    text_parts: list[str] = []

    for block in content:
        block_type = block.get("type") if isinstance(block, dict) else getattr(block, "type", None)

        if block_type == "tool_use":
            block_id = block.get("id") if isinstance(block, dict) else getattr(block, "id", "")
            block_name = block.get("name") if isinstance(block, dict) else getattr(block, "name", "")
            block_input = block.get("input") if isinstance(block, dict) else getattr(block, "input", {})

            tool_calls.append(
                ToolCall(
                    id=block_id,
                    name=block_name,
                    arguments=json.dumps(block_input, separators=(",", ":"))
                    if not isinstance(block_input, str)
                    else block_input,
                )
            )
        elif block_type == "text":
            block_text = block.get("text") if isinstance(block, dict) else getattr(block, "text", "")
            text_parts.append(block_text)

    # If tool calls present, return them (pipeline handles the loop)
    if tool_calls:
        return tool_calls

    # Text content
    text = "".join(text_parts)
    if not text:
        return None

    # Structured output — JSON parse when outputs schema exists
    outputs = getattr(agent, "outputs", None)
    if outputs:
        try:
            return json.loads(text)
        except (json.JSONDecodeError, ValueError):
            return text

    return text


# ---------------------------------------------------------------------------
# Streaming
# ---------------------------------------------------------------------------


def _stream_generator(response: Iterator) -> Iterator[str | Any]:
    """Yield content chunks from an Anthropic streaming response.

    Handles streaming event types:
    - ``content_block_delta`` with ``delta.type == "text_delta"`` → yield text
    - ``content_block_start`` with ``content_block.type == "tool_use"`` → accumulate tool call
    - ``input_json_delta`` → accumulate partial JSON for tool arguments
    - Tool calls are yielded at the end of the stream.
    """
    from ..openai.processor import ToolCall

    tool_call_acc: dict[int, dict[str, str]] = {}

    for event in response:
        e = event if isinstance(event, dict) else getattr(event, "__dict__", {})
        event_type = e.get("type") if isinstance(e, dict) else getattr(event, "type", None)

        if event_type == "content_block_delta":
            delta = e.get("delta") if isinstance(e, dict) else getattr(event, "delta", None)
            if not delta:
                continue
            delta_type = delta.get("type") if isinstance(delta, dict) else getattr(delta, "type", None)

            if delta_type == "text_delta":
                text = delta.get("text") if isinstance(delta, dict) else getattr(delta, "text", "")
                yield text
            elif delta_type == "input_json_delta":
                idx = e.get("index") if isinstance(e, dict) else getattr(event, "index", 0)
                acc = tool_call_acc.get(idx)
                if acc:
                    partial = (
                        delta.get("partial_json", "") if isinstance(delta, dict) else getattr(delta, "partial_json", "")
                    )
                    acc["arguments"] += partial

        elif event_type == "content_block_start":
            block = e.get("content_block") if isinstance(e, dict) else getattr(event, "content_block", None)
            if not block:
                continue
            block_type = block.get("type") if isinstance(block, dict) else getattr(block, "type", None)
            if block_type == "tool_use":
                idx = e.get("index") if isinstance(e, dict) else getattr(event, "index", 0)
                block_id = block.get("id", "") if isinstance(block, dict) else getattr(block, "id", "")
                block_name = block.get("name", "") if isinstance(block, dict) else getattr(block, "name", "")
                tool_call_acc[idx] = {"id": block_id, "name": block_name, "arguments": ""}

    # Yield accumulated tool calls at end of stream
    for idx in sorted(tool_call_acc):
        tc = tool_call_acc[idx]
        yield ToolCall(id=tc["id"], name=tc["name"], arguments=tc["arguments"])


async def _async_stream_generator(response: AsyncIterator) -> AsyncIterator[str | Any]:
    """Async variant of :func:`_stream_generator`."""
    from ..openai.processor import ToolCall

    tool_call_acc: dict[int, dict[str, str]] = {}

    async for event in response:
        e = event if isinstance(event, dict) else getattr(event, "__dict__", {})
        event_type = e.get("type") if isinstance(e, dict) else getattr(event, "type", None)

        if event_type == "content_block_delta":
            delta = e.get("delta") if isinstance(e, dict) else getattr(event, "delta", None)
            if not delta:
                continue
            delta_type = delta.get("type") if isinstance(delta, dict) else getattr(delta, "type", None)

            if delta_type == "text_delta":
                text = delta.get("text") if isinstance(delta, dict) else getattr(delta, "text", "")
                yield text
            elif delta_type == "input_json_delta":
                idx = e.get("index") if isinstance(e, dict) else getattr(event, "index", 0)
                acc = tool_call_acc.get(idx)
                if acc:
                    partial = (
                        delta.get("partial_json", "") if isinstance(delta, dict) else getattr(delta, "partial_json", "")
                    )
                    acc["arguments"] += partial

        elif event_type == "content_block_start":
            block = e.get("content_block") if isinstance(e, dict) else getattr(event, "content_block", None)
            if not block:
                continue
            block_type = block.get("type") if isinstance(block, dict) else getattr(block, "type", None)
            if block_type == "tool_use":
                idx = e.get("index") if isinstance(e, dict) else getattr(event, "index", 0)
                block_id = block.get("id", "") if isinstance(block, dict) else getattr(block, "id", "")
                block_name = block.get("name", "") if isinstance(block, dict) else getattr(block, "name", "")
                tool_call_acc[idx] = {"id": block_id, "name": block_name, "arguments": ""}

    for idx in sorted(tool_call_acc):
        tc = tool_call_acc[idx]
        yield ToolCall(id=tc["id"], name=tc["name"], arguments=tc["arguments"])
